convert/MaskCode.ahk

```diff
@@ -0,0 +1,435 @@
+
+	; acts as constructor for a node object
+	__New(name, cType, pos, len)
+	{
+		this.name			:= name
+		this.cType			:= cType
+		this.pos			:= pos
+		this.len			:= len
+	}
+
+	; PRIVATE - returns the nested path depth of the node
+	_depth() {
+		StrReplace(this.PathVal, "\",,, &count)
+		return count
+	}
+
+;	; PRIVATE - custom sort for path depth
+;	_depthSort(a1,a2,*)
+;	{
+;		RegExMatch(a1, "(\d+):(\d+):", &m1)
+;		RegExMatch(a2, "(\d+):(\d+):", &m2)
+;		return ((m2[2] > m1[2]) ? 1 : ((m2[2] < m1[2]) ? -1 : ((m2[1] > m1[1]) ? 1 : ((m2[1] < m1[1]) ? -1 : 0))))
+;	}
+
+	; PUBLIC - convenience - returns string list of ChildList map()
+	GetChildren() {
+		cList := ""
+		for key, childNode in this.ChildList {
+			cList .= ((cList="") ? "" : ";") . ("[" key "]" . childNode.name)
+		}
+		return cList
+	}
+
+	EndPos					=> this.pos + this.len
+	ParentName				=> (this.parentPos > 0)	 ? NodeMap.mapList[this.parentPos].name : "Root"
+	Path					=> ((this.parentPos > 0) ? NodeMap.mapList[this.parentPos].Path : "") . ("\" this.name)
+	PathVal					=> ((this.parentPos > 0) ? NodeMap.mapList[this.parentPos].PathVal : "") . ("\" this.pos)
+	AddChild(id)			=> this.ChildList[id] := NodeMap.mapList[id]	; add node object
+	hasChildren				=> this.ChildList.Count
+
+	;################################################################################
+
+	static getNode(id)		=> NodeMap.mapList(id)
+	static getName(id)		=> NodeMap.mapList(id).name
+	static mapList			:= map()
+	static idIndex			:= 0
+	static nextIdx			=> ++NodeMap.IdIndex
+
+	; PRIVATE - adds a node to maplist
+	static _add(node) {
+		NodeMap.mapList[node.pos] := node
+		return
+	}
+
+	; PUBLIC - provides details of all nodes in maplist
+	; used for debugging, etc.
+	static Report() {
+		reportStr := ""
+		for key, nm in NodeMap.MapList {
+			reportStr	.= "`nname:`t[" nm.cType "] " nm.name "`nstart:`t" nm.pos "`nend:`t" nm.EndPos "`nlen:`t" nm.len
+						. "`nparent:`t" nm.parentName " [" nm.parentPos "]`npath:`t" nm.path "`npathV:`t" nm.pathVal "`nDepth:`t" nm.Depth()
+						. "`nPList:`t" nm.parentList "`nChilds:`t" nm.getChildren() "`n"
+		}
+		return reportStr
+	}
+
+	; PUBLIC - builds a position map of all classes and functions found in script
+	;	also identifies relationship between nodes
+	static BuildNodeMap(code)
+	{
+		NodeMap.Reset()		; each build requires a fresh MapList
+
+		; map all classes - including nested ones, from top to bottom
+		pos := 0
+		while(pos := RegExMatch(code, gClassPtn, &m, pos+1)) {
+			NodeMap._add(NodeMap(m.cname, "CLS", pos, m.len))
+		}
+
+		; map all functions - including nested ones, from top to bottom
+		pos := 0
+		while(pos := RegExMatch(code, gFuncPtn, &m, pos+1)) {
+			if (m[]="")
+				continue	; bypass IF/WHILE/LOOP
+			NodeMap._add(NodeMap(m.fname, "FUNC", pos, m.len))
+		}
+
+		; identify parents and children for each node in maplist
+		NodeMap._setKin()
+		return
+	}
+
+	; PUBLIC - mask and convert classes and functions
+	static MaskAndConvertNodes(&code)
```

Comment:
So does existing code to convert functions eg `subLoopFunctions()` still get used?

Reply:
> So does existing code to convert functions eg `subLoopFunctions()` still get used?

Yes, the conversions are still handled by the exact same function as before (looping the lines). That function is called by the new Convert() func directly above it. However, prior to sending the code to that function, the classes and functions are masked so that the _convertLines function only receives the main body (global code) of the script. The classes and functions are essentially removed prior to _convertLines receives the code.

The classes and functions are basically extracted out of the original code, then processed separately. Each class/function block becomes an individual/single unit. Nested blocks (parent/child) are also isolated from each other and handled separately. So basically no function or class knows about each other during the conversion of those blocks. They are disconnected so that their variable scope does not cross borders. 

Now, this can cause issues in some cases... when a variable is expected to be read/manipulated in the parent OR child block. I don't know what the percentage of v1 cases are using nested functions, but I'm sure it is less often than when you have multiple separate functions the same level/depth.

Tweaks may need to be made to the masking code when it causes issues. Testing wiil be required to see whether global separation is more/less desirable than isolated cases as needed. Both can be supported, but this initial submission is focused mostly on an masking from a global perspective (all code rather than isolated cases).

The finalize code was there from the beginning at the end of the convert function. But it caused issues so I enclosed it in an IF block for now until a more appropriate solution can be created. The current solution is still a work in progress but I wanted to get the masking code out there to be tested and evaluated by others. I can't anticipate every possible scenario, so getting feedback is the best way to proceed with "perfecting" the idea.

I'm sure tweaks will need to be made and the current configuration is not set in stone. Just an initial draft.

Andy

Reply:
Ahh got it, couldn't find any other issues so I'll merge